File: frontends/civil-officer/src/situationFamiliale.ts
import type { Sexe } from "./registry";

export type CoteFamille = "HOMME" | "FEMME" | "PERSONNE" | "";

export type FamilyMember = {
  nom: string;
  postnom: string;
  prenom: string;
  sexe: Sexe | "";
  date_naissance: string;
  telephone: string;
  vit_avec: boolean;
  lien: string;
  cote: CoteFamille;
  person_id?: string | null;
};

export type SituationFamilialeData = {
  a_conjoint: boolean;
  conjoint: FamilyMember;
  nombre_enfants: number;
  enfants: FamilyMember[];
  personnes_a_charge: FamilyMember[];
  remarques: string;
};

export const COTES_FAMILLE: { value: CoteFamille; label: string }[] = [
  { value: "", label: "—" },
  { value: "HOMME", label: "Côté homme (époux)" },
  { value: "FEMME", label: "Côté femme (épouse)" },
  { value: "PERSONNE", label: "Côté de la personne recensée" },
];

export const LIENS_PARENTE = [
  "",
  "Enfant",
  "Petit frère",
  "Petite sœur",
  "Frère",
  "Sœur",
  "Neveu",
  "Nièce",
  "Oncle",
  "Tante",
  "Cousin",
  "Cousine",
  "Beau-père",
  "Belle-mère",
  "Beau-frère",
  "Belle-sœur",
  "Petit-fils",
  "Petite-fille",
  "Grand-père",
  "Grand-mère",
  "Travailleur / employé(e)",
  "Autre personne à charge",
] as const;

export function emptyMember(lien = "", cote: CoteFamille = ""): FamilyMember {
  return {
    nom: "",
    postnom: "",
    prenom: "",
    sexe: "",
    date_naissance: "",
    telephone: "",
    vit_avec: true,
    lien,
    cote,
    person_id: null,
  };
}

export function emptySituationFamiliale(): SituationFamilialeData {
  return {
    a_conjoint: false,
    conjoint: emptyMember("CONJOINT"),
    nombre_enfants: 0,
    enfants: [],
    personnes_a_charge: [],
    remarques: "",
  };
}

/** Ajuste la liste des enfants au nombre choisi (0–5 cliquable ; >5 possible). */
export function resizeEnfants(current: FamilyMember[], count: number): FamilyMember[] {
  const n = Math.max(0, Math.floor(count));
  if (current.length === n) return current;
  if (current.length > n) return current.slice(0, n);
  const next = [...current];
  while (next.length < n) next.push(emptyMember("Enfant"));
  return next;
}

function coteLabel(c: CoteFamille): string {
  return COTES_FAMILLE.find((x) => x.value === c)?.label ?? "";
}

function memberLabel(m: FamilyMember): string {
  const name = [m.nom, m.postnom, m.prenom].map((x) => x.trim()).filter(Boolean).join(" ");
  const bits = [name || "—"];
  if (m.sexe) bits.push(m.sexe === "M" ? "M" : "F");
  if (m.date_naissance) bits.push(`né(e) ${m.date_naissance}`);
  if (m.telephone.trim()) bits.push(`tél. ${m.telephone.trim()}`);
  if (m.cote) bits.push(coteLabel(m.cote));
  if (m.lien.trim()) bits.push(`lien: ${m.lien.trim()}`);
  if (!m.vit_avec) bits.push("ne vit pas avec");
  return bits.join(", ");
}

export function formatSituationFamiliale(data: SituationFamilialeData): string {
  const lines: string[] = [];
  if (data.a_conjoint) {
    lines.push(`Conjoint(e) : ${memberLabel(data.conjoint)}`);
  } else {
    lines.push("Conjoint(e) : aucun");
  }
  lines.push(`Nombre d'enfants : ${data.nombre_enfants}`);
  if (data.enfants.length) {
    lines.push(`Enfants (${data.enfants.length}) :`);
    data.enfants.forEach((e, i) => lines.push(`  ${i + 1}. ${memberLabel(e)}`));
  } else {
    lines.push("Enfants : aucun déclaré");
  }
  if (data.personnes_a_charge.length) {
    lines.push(`Personnes à charge (${data.personnes_a_charge.length}) :`);
    data.personnes_a_charge.forEach((p, i) => lines.push(`  ${i + 1}. ${memberLabel(p)}`));
  } else {
    lines.push("Personnes à charge : aucune");
  }
  if (data.remarques.trim()) {
    lines.push(`Remarques : ${data.remarques.trim()}`);
  }
  return lines.join("\n");
}

export function parseSituationFamiliale(raw: unknown): SituationFamilialeData {
  const base = emptySituationFamiliale();
  if (!raw || typeof raw !== "object") {
    if (typeof raw === "string" && raw.trim()) {
      return { ...base, remarques: raw.trim() };
    }
    return base;
  }
  const d = raw as Partial<SituationFamilialeData>;
  const enfants = Array.isArray(d.enfants)
    ? d.enfants.map((e) => ({ ...emptyMember("Enfant"), ...e }))
    : [];
  const nombre =
    typeof d.nombre_enfants === "number" && d.nombre_enfants >= 0
      ? d.nombre_enfants
      : enfants.length;
  return {
    a_conjoint: Boolean(d.a_conjoint),
    conjoint: { ...emptyMember("CONJOINT"), ...(d.conjoint ?? {}) },
    nombre_enfants: nombre,
    enfants,
    personnes_a_charge: Array.isArray(d.personnes_a_charge)
      ? d.personnes_a_charge.map((e) => ({ ...emptyMember(""), ...e }))
      : [],
    remarques: typeof d.remarques === "string" ? d.remarques : "",
  };
}

File: apps/flutter_recensement/lib/features/census/situation_familiale.dart
import 'dart:convert';

class FamilyMember {
  FamilyMember({
    this.nom = '',
    this.postnom = '',
    this.prenom = '',
    this.sexe = '',
    this.dateNaissance = '',
    this.telephone = '',
    this.vitAvec = true,
    this.lien = '',
  });

  String nom;
  String postnom;
  String prenom;
  String sexe;
  String dateNaissance;
  String telephone;
  bool vitAvec;
  String lien;

  Map<String, dynamic> toJson() => {
        'nom': nom,
        'postnom': postnom,
        'prenom': prenom,
        'sexe': sexe,
        'date_naissance': dateNaissance,
        'telephone': telephone,
        'vit_avec': vitAvec,
        'lien': lien,
      };

  static FamilyMember fromJson(Map<String, dynamic>? raw, {String lien = ''}) {
    if (raw == null) return FamilyMember(lien: lien);
    return FamilyMember(
      nom: raw['nom']?.toString() ?? '',
      postnom: raw['postnom']?.toString() ?? '',
      prenom: raw['prenom']?.toString() ?? '',
      sexe: raw['sexe']?.toString() ?? '',
      dateNaissance: raw['date_naissance']?.toString() ?? '',
      telephone: raw['telephone']?.toString() ?? '',
      vitAvec: raw['vit_avec'] != false,
      lien: raw['lien']?.toString() ?? lien,
    );
  }

  String get label {
    final name = [nom, postnom, prenom].map((e) => e.trim()).where((e) => e.isNotEmpty).join(' ');
    final bits = <String>[name.isEmpty ? '—' : name];
    if (sexe.isNotEmpty) bits.add(sexe == 'M' ? 'M' : 'F');
    if (dateNaissance.isNotEmpty) bits.add('né(e) $dateNaissance');
    if (telephone.trim().isNotEmpty) bits.add('tél. ${telephone.trim()}');
    if (!vitAvec) bits.add('ne vit pas avec');
    if (lien.trim().isNotEmpty) bits.add('lien: ${lien.trim()}');
    return bits.join(', ');
  }
}

class SituationFamiliale {
  SituationFamiliale({
    this.aConjoint = false,
    FamilyMember? conjoint,
    List<FamilyMember>? enfants,
    List<FamilyMember>? personnesACharge,
    this.remarques = '',
  })  : conjoint = conjoint ?? FamilyMember(lien: 'CONJOINT'),
        enfants = enfants ?? <FamilyMember>[],
        personnesACharge = personnesACharge ?? <FamilyMember>[];

  bool aConjoint;
  FamilyMember conjoint;
  List<FamilyMember> enfants;
  List<FamilyMember> personnesACharge;
  String remarques;

  Map<String, dynamic> toJson() => {
        'a_conjoint': aConjoint,
        'conjoint': conjoint.toJson(),
        'enfants': enfants.map((e) => e.toJson()).toList(),
        'personnes_a_charge': personnesACharge.map((e) => e.toJson()).toList(),
        'remarques': remarques,
      };

  String formatSummary() {
    final lines = <String>[];
    lines.add(aConjoint ? 'Conjoint(e) : ${conjoint.label}' : 'Conjoint(e) : aucun');
    if (enfants.isEmpty) {
      lines.add('Enfants : aucun déclaré');
    } else {
      lines.add('Enfants (${enfants.length}) :');
      for (var i = 0; i < enfants.length; i++) {
        lines.add('  ${i + 1}. ${enfants[i].label}');
      }
    }
    if (personnesACharge.isEmpty) {
      lines.add('Personnes à charge : aucune');
    } else {
      lines.add('Personnes à charge (${personnesACharge.length}) :');
      for (var i = 0; i < personnesACharge.length; i++) {
        lines.add('  ${i + 1}. ${personnesACharge[i].label}');
      }
    }
    if (remarques.trim().isNotEmpty) {
      lines.add('Remarques : ${remarques.trim()}');
    }
    return lines.join('\n');
  }

  static SituationFamiliale parse(Object? raw) {
    if (raw == null) return SituationFamiliale();
    if (raw is String) {
      final t = raw.trim();
      if (t.isEmpty) return SituationFamiliale();
      try {
        final decoded = jsonDecode(t);
        if (decoded is Map) return parse(decoded);
      } catch (_) {}
      return SituationFamiliale(remarques: t);
    }
    if (raw is! Map) return SituationFamiliale();
    final m = Map<String, dynamic>.from(raw);
    final enfantsRaw = m['enfants'];
    final chargesRaw = m['personnes_a_charge'];
    return SituationFamiliale(
      aConjoint: m['a_conjoint'] == true,
      conjoint: FamilyMember.fromJson(
        m['conjoint'] is Map ? Map<String, dynamic>.from(m['conjoint'] as Map) : null,
        lien: 'CONJOINT',
      ),
      enfants: enfantsRaw is List
          ? enfantsRaw
              .whereType<Map>()
              .map((e) => FamilyMember.fromJson(Map<String, dynamic>.from(e), lien: 'ENFANT'))
              .toList()
          : <FamilyMember>[],
      personnesACharge: chargesRaw is List
          ? chargesRaw
              .whereType<Map>()
              .map((e) => FamilyMember.fromJson(Map<String, dynamic>.from(e)))
              .toList()
          : <FamilyMember>[],
      remarques: m['remarques']?.toString() ?? '',
    );
  }
}
